### PR TITLE
Add DynamicMoE support for Mixtral (#10)

### PR DESCRIPTION
**Pull request by astachowiczhabana**
Add DynamicMoE support for Mixtral

### CODE REVIEW

**Comment by astachowiczhabana** [comment]
Hi @libinta this commit is also required with next OH release

**Comment by jiminha** [comment]
@astachowiczhabana  this is dupe with  PR 1511? https://github.com/huggingface/optimum-habana/pull/1511

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 1]
```diff
@@ -449,6 +450,172 @@ def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) ->
...
+        routing_weights, selected_experts = calculate_routing_tensors(score, topk, hidden_states.dtype)
+        # pre-processing for custom op inputs
+        experts_range = range(self.num_total_experts)
+        w13_list = [w13[i, :, :].squeeze() for i in experts_range]
+        w2_list = [w2[i, :, :].squeeze() for i in experts_range]
+        final_hidden_states = torch.ops.hpu.mixture_of_experts(
+            hidden_states=hidden_states,
+            expert_routing_table=selected_experts,
+            router_weights=routing_weights,
+            w12=w13_list,
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:474]: do you have a comment on this quote?  https://github.com/huggingface/optimum-habana/pull/1466#issuecomment-2454643810 
"According to the conversation with Ilay, it is not required to manually fuse w1 and w3 into one weight, and **keeping them separately is better than merging them in the case of fine-tuning** considering in 1.19 Synapse release the Custom MoE kernel will support training."

File: optimum/habana/checkpoint_utils.py [hunk 2]
```diff
@@ -94,7 +94,7 @@ def model_on_meta(config):
     """
     Checks if load the model to meta.
     """
-    return config.model_type in ["bloom", "llama", "falcon", "mixtral", "qwen2"]
+    return config.model_type in ["bloom", "llama", "falcon", "qwen2"]
```
Comment (schoi-habana) [inline optimum/habana/checkpoint_utils.py:97]: i guess this removal is due to the w1 and w3 merge when loading the model from ckpt. but that is irrelevant to static moe and static moe can still have benefit from loading to meta. anyway to separate those 2 cases?

File: optimum/habana/transformers/models/mixtral/modeling_mixtral.py [hunk 3]
```diff
@@ -449,6 +450,172 @@ def gaudi_mixtral_block_sparse_moe_forward(self, hidden_states: torch.Tensor) ->
...
+            w3=w2_list,
+            permuted_weights=True,
+            activation="silu",
+            experts_min=0,
+            experts_max=7,
+        )
+        return final_hidden_states.view(-1, hidden_states.shape[1])
+
+
+def calculate_routing_tensors(
```
Comment (schoi-habana) [inline optimum/habana/transformers/models/mixtral/modeling_mixtral.py:484]: this is common for both static and dynamic moe. can it be reused in static moe?

**Pull request closed without merging**